### ISSUE TITLE
Changing development server port does not take effect

### ISSUE THREAD

**Issue by btoconnor** [rails/rails#27834]
Sorry if this is a bad issue - I'm just starting out with rails and encountered this.  I did a search on GitHub and didn't find anything relating to this issue I'm seeing.   Please let me know if I messed something up or you need more information.

### Steps to reproduce
This is on a fresh install - right from the guide.  I ran the following commands:

```
rails new blog
cd blog
./bin/rails server -p 3002
```
And this is the output:
```
/home/brian/.rvm/gems/ruby-2.4.0-rc1/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:51: warning: constant ::Fixnum is deprecated
/home/brian/.rvm/gems/ruby-2.4.0-rc1/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:52: warning: constant ::Bignum is deprecated
=> Booting Puma
=> Rails 5.0.1 application starting in development on http://localhost:3002
=> Run `rails server -h` for more startup options
/home/brian/.rvm/gems/ruby-2.4.0-rc1/gems/activesupport-5.0.1/lib/active_support/core_ext/numeric/conversions.rb:138: warning: constant ::Fixnum is deprecated
Puma starting in single mode...
* Version 3.7.0 (ruby 2.4.0-p-1), codename: Snowy Sagebrush
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
Exiting
```

### Expected behavior
The server should attempt to start on port 3002.

### Actual behavior
The server attempts to start on port 3000.  As you can see in the logs above, the rails server acknowledges the command line parameter (in the log line "Rails 5.0.1 application starting in development on http://localhost:3002"), but below it attempts to listen on port 3000 ("Listening on tcp://0.0.0.0:3000")

### System configuration
**Rails version**: Rails 5.0.1

**Ruby version**: ruby 2.4.0rc1 (2016-12-12 trunk 57064) [x86_64-linux]

**Comment by Fukumasuta** [comment]
I think you just have to do the following :+1: `rails s -p 3002`
if you are on the root folder as it seems to be. For me, it works.

**Comment by btoconnor** [comment]
Yeah - that doesn't change anything.  Note how the two debug lines are different - the initial one thinks it's going to port 3002, whereas the actual binding is happening on 3000.

For posterity, I tried it with your command as well though:


```
brian@kansas:~/Programming/rails/blog$ rails s -p 3002
/home/brian/.rvm/gems/ruby-2.4.0-rc1/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:51: warning: constant ::Fixnum is deprecated
/home/brian/.rvm/gems/ruby-2.4.0-rc1/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:52: warning: constant ::Bignum is deprecated
=> Booting Puma
=> Rails 5.0.1 application starting in development on http://localhost:3002
=> Run `rails server -h` for more startup options
/home/brian/.rvm/gems/ruby-2.4.0-rc1/gems/activesupport-5.0.1/lib/active_support/core_ext/numeric/conversions.rb:138: warning: constant ::Fixnum is deprecated
Puma starting in single mode...
* Version 3.7.0 (ruby 2.4.0-p-1), codename: Snowy Sagebrush
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
```

Note again how it's binding to port 3000, despite passing in 3002

**Comment by Fukumasuta** [comment]
I see...
Well, you seems to have a RC (release candidate) version of Ruby and those two lines : 
`/home/brian/.rvm/gems/ruby-2.4.0-rc1/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:51: warning: constant ::Fixnum is deprecated
/home/brian/.rvm/gems/ruby-2.4.0-rc1/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:52: warning: constant ::Bignum is deprecated`
does not showed up from me. 

You may try to install a standard version of ruby like ruby 2.3.0 (i'm using this version without any problem). Though, i cannot guarantee this is the real solution...

**Comment by btoconnor** [comment]
Interesting.  I ended up installing ruby 2.3 to test out your theory, but it seems like it didn't solve it.  It got rid of the warnings, but it's still binding on the wrong port:
```
brian@kansas:~/Programming/rails/blog$ rails s -p 3002
=> Booting Puma
=> Rails 5.0.1 application starting in development on http://localhost:3002
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.7.0 (ruby 2.3.3-p222), codename: Snowy Sagebrush
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
Use Ctrl-C to stop
```

**Comment by ayghor** [comment]
i can change ports without problem on rails 5.0.1 development.

our outputs says:

    http://localhost:3002

but yours:

    tcp://0.0.0.0:3000

and mine:

    tcp://localhost:3002

which differs not only on port, but on bound ip addresses. `localhost` will allow connection only from `localhost`, whiel `0.0.0.0` will allow connections from anywhere AFAIK.

you could try to bind to another ips liek `rails -b 127.0.0.1` or `rails -b 1.2.3.4` and in combination with teh `-p 3002` to see what it outputs.

=O3

**Comment by btoconnor** [comment]
Looks like same thing:

```
brian@kansas:~/Programming/rails/blog$ rails s -b 127.0.0.1 -p 3002
=> Booting Puma
=> Rails 5.0.1 application starting in development on http://127.0.0.1:3002
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.7.0 (ruby 2.3.3-p222), codename: Snowy Sagebrush
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
Use Ctrl-C to stop
```

It seems to me that the defaults when passing from the rails server to the Rack server aren't being preserved (sorry if I butchered that terminology somehow) - but that's just my guess (as I mentioned, newcomer to ruby and rails.

Happy to provide any more information if that's helpful.

I just installed everything on this computer today

**Comment by ayghor** [comment]
maybe using another server? `unicorn`, `thin`, ... ?

**Comment by btoconnor** [comment]
Sure, I could.  For what it's worth - I found this because I had another server running on port 3000.  I stopped the server so that I could run through the tutorial, but it's not really the point.

If the development server isn't listening to the arguments, it seems like a bug.  I can certainly work around it, but it seems like it should be fixed, which is why I filed the issue.

**Comment by ayghor** [comment]
yeaa, but it sure makes a difference if teh bug affects only puma and not others.

**Comment by ayghor** [comment]
btw, did you try changing ports on production environment too?

**Comment by AaronCowan** [comment]
@btoconnor if you want to change the port that the puma server is running you would have to to go into the `config/puma.rb` file and change 

`port        ENV.fetch("PORT") { 3000 }`
to
`port        ENV.fetch("PORT") { 3002 }`

**Comment by btoconnor** [comment]
I was able to launch it with webrick on the correct port:


```
brian@kansas:~/Programming/rails/blog$ rails s -b 127.0.0.1 -p 3002 webrick
=> Booting WEBrick
=> Rails 5.0.1 application starting in development on http://127.0.0.1:3002
=> Run `rails server -h` for more startup options
[2017-01-28 12:36:48] INFO  WEBrick 1.3.1
[2017-01-28 12:36:48] INFO  ruby 2.3.3 (2016-11-21) [x86_64-linux]
[2017-01-28 12:36:48] INFO  WEBrick::HTTPServer#start: pid=23590 port=3002
```

@AaronCowan - I can modify that.  I just think that if rails is using puma as the default server, and it thinks it's going to launch the server on a port, and it doesn't, that's a bug.  It clearly shows the log line saying it thinks it's going to launch on 3002, and the underlying server doesn't.  If it's not going to override on the command line, it should just throw an error if you try to modify the port / host through the cli telling you to modify the config instead.  Either way, it's confusing, and it should be made clearer, which I think is a bug.

**Comment by ayghor** [comment]
and if you start it liek:

  PORT=3002 rails s

 does teh port change to 3002 or stay 3000?

**Comment by btoconnor** [comment]
That does appear to work:

```
brian@kansas:~/Programming/rails/blog$ PORT=3002 rails s -b 127.0.0.1
=> Booting Puma
=> Rails 5.0.1 application starting in development on http://127.0.0.1:3002
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.7.0 (ruby 2.3.3-p222), codename: Snowy Sagebrush
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3002
```

**Comment by AaronCowan** [comment]
Would someone let me know if the following solution is any good, if so I would submit a pull request.

- At start time check to see if the server is Puma
- If the server is Puma set ENV["PORT"] = options[:Port]

The code change would be in `/rails/railties/lib/rails/commands/server/server_command.rb`

```ruby
def initialize(options = nil)
  @default_options = options || {}
  super(@default_options)
  set_port if puma_server?
  set_environment
end

private
  def puma_server?
    ActiveSupport::Inflector.demodulize(server) == 'Puma'
  end

  def set_port
    ENV["PORT"] = options[:port]
  end
```

**Comment by btoconnor** [comment]
I can't comment much on the code itself, but I will say that it appears as if the host isn't being set too.  So, if you're fixing the port, it's probably good to fix the host as well.

Also, thanks for taking a look at this.

**Comment by rafaelfranca** [comment]
Rails should not have any code specific for Puma. This is a bug in Puma
that is not respecting the rack port that is being set.

**Comment by y-yagi** [comment]
Duplicate of #24435

**Comment by rafaelfranca** [comment]
It seems to be a bug with puma 3.7.0. Could you open an issue in the puma issue tracker?

**Comment by ayghor** [comment]
see https://github.com/puma/puma/commit/42bec4600c51ab8a1c1ee5a0e1b738a4ffd82bf2, it was tagged `v3.6.2` and was __NOT__ merged on `v3.7.0` for some reason:

>     Revert "fix access priorities of each level in LeveledOptions"
>     
>     This reverts commit 15f0847289d0815284f87eeea4625100b581c026.
>     
>     See #1154, this commit unintentionally changed behavior regarding the
>     importance of command line options.

cant test reverting teh reversion right now
